# Practice Loading and Exploring Datasets

This assignment is purposely open-ended. You will be asked to load datasets from the [UC-Irvine Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php). 

Even though you maybe using different datasets than your fellow classmates, try and be supportive and assist each other in the challenges that you are facing. You will only deepend your understanding of these topics as you work to assist one another. Many popular UCI datasets face similar data loading challenges.

Remember that the UCI datasets do not necessarily have a file type of `.csv` so it's important that you learn as much as you can about the dataset before you try and load it. See if you can look at the raw text of the file either locally or using the `!curl` shell command -or in some other way before you try and read it in as a dataframe. This will help you catch what would otherwise be unforseen problems.

Feel free to embellish this notebook with additional markdown cells,code cells, comments, graphs, etc. Whatever you think helps adequately address the questions.

## 1) Load a dataset from UCI (via its URL)

Please navigate to the home page and choose a dataset (other than the Adult dataset) from the "Most Popular" section on the right-hand side of the home page. Load the dataset via its URL and check the following (show your work):

- Are the headers showing up properly?
- Look at the first 5 and the last 5 rows, do they seem to be in order?
- Does the dataset have the correct number of rows and columns as described in the UCI page? 
 - Remember, that UCI does not count the y variable (column of values that we might want to predict via a machine learning model) as an "attribute" but rather as a "class attribute" so you may end up seeing a number of columns that is one greater than the number listed on the UCI website.
- Does UCI list this dataset as having missing values? Check for missing values and see if your analysis corroborates what UCI reports?
- if `NaN` values or other missing value indicators are not being detected by `df.isnull().sum()` find a way to replace whatever is indicating the missing values with `np.NaN`.
- Use the .describe() function in order to see the summary statistics of both the numeric and non-numeric columns. 

In [13]:
#Generic header importing certain libraries as shortform to make coding nicer. 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#storing the URL of the dataset prior to sanitizing the data. 
wine_data_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv'

#starting a pandas dataframe -- which will assist in fall futher analyses. 
wine_df = pd.read_csv(wine_data_url, sep=";")

#looking at the first 5 entries of the dataset 
wine_df.head(5)

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0            7.4              0.70         0.00  ...       0.56      9.4        5
1            7.8              0.88         0.00  ...       0.68      9.8        5
2            7.8              0.76         0.04  ...       0.65      9.8        5
3           11.2              0.28         0.56  ...       0.58      9.8        6
4            7.4              0.70         0.00  ...       0.56      9.4        5

[5 rows x 12 columns]

In [16]:
#looking at the last 5 entries of the dataset 
wine_df.tail(5)

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
1594            6.2             0.600         0.08  ...       0.58     10.5        5
1595            5.9             0.550         0.10  ...       0.76     11.2        6
1596            6.3             0.510         0.13  ...       0.75     11.0        6
1597            5.9             0.645         0.12  ...       0.71     10.2        5
1598            6.0             0.310         0.47  ...       0.66     11.0        6

[5 rows x 12 columns]

In [17]:
#high level statistics on the dataset
wine_df.describe()

fixed acidity  volatile acidity  ...      alcohol      quality
count    1599.000000       1599.000000  ...  1599.000000  1599.000000
mean        8.319637          0.527821  ...    10.422983     5.636023
std         1.741096          0.179060  ...     1.065668     0.807569
min         4.600000          0.120000  ...     8.400000     3.000000
25%         7.100000          0.390000  ...     9.500000     5.000000
50%         7.900000          0.520000  ...    10.200000     6.000000
75%         9.200000          0.640000  ...    11.100000     6.000000
max        15.900000          1.580000  ...    14.900000     8.000000

[8 rows x 12 columns]

In [19]:
#looking at the number of rows and columns. 
wine_df.shape

(1599, 12)

In [20]:
#looking at the index, datatype, and memory information
wine_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64
chlorides               1599 non-null float64
free sulfur dioxide     1599 non-null float64
total sulfur dioxide    1599 non-null float64
density                 1599 non-null float64
pH                      1599 non-null float64
sulphates               1599 non-null float64
alcohol                 1599 non-null float64
quality                 1599 non-null int64
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [9]:
#probing to see if there is any missing values (NaNs)
wine_df.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [21]:
wine_df.count()

fixed acidity           1599
volatile acidity        1599
citric acid             1599
residual sugar          1599
chlorides               1599
free sulfur dioxide     1599
total sulfur dioxide    1599
density                 1599
pH                      1599
sulphates               1599
alcohol                 1599
quality                 1599
dtype: int64

In [26]:
wine_df.notnull().sum()

fixed acidity           1599
volatile acidity        1599
citric acid             1599
residual sugar          1599
chlorides               1599
free sulfur dioxide     1599
total sulfur dioxide    1599
density                 1599
pH                      1599
sulphates               1599
alcohol                 1599
quality                 1599
dtype: int64

## Answers to Question Set 1: 

1. Yes the headers are showing up nicely as we can see in the earlier cells when you execude the .head() you can see some values showing up -- relevantly clean. 

2. Looking at one of the prior cells where the first and last five values were read -- everything seems in order

3. Looking at the UCI info section for the RED wine dataset we can see that there are supposed to be 1599 instances (rows)-- based on tail we see we have 1598, this is due to indexing at 0; so in total we have 1599 instances. There are supposed be 12 variables (columns) on which we have data -- and there are 12 columns in the dataset. 

4. The values of minimums and maximums seem to make sense -- like acidity and % values... etc.  

## 2) Load a dataset from your local machine.
Choose a second dataset from the "Popular Datasets" listing on UCI, but this time download it to your local machine instead of reading it in via the URL. Upload the file to Google Colab using the files tab in the left-hand sidebar or by importing `files` from `google.colab` The following link will be a useful resource if you can't remember the syntax: <https://towardsdatascience.com/3-ways-to-load-csv-files-into-colab-7c14fcbdcb92>

- Answer all of the same bullet point questions from part 1 again on this new dataset. 


In [32]:
#datasets chosen -- https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29

"""
column information for wpbc.ndata 
1) ID number
2) Diagnosis (M = malignant, B = benign)
3-32)

Ten real-valued features are computed for each cell nucleus:

a) radius (mean of distances from center to points on the perimeter)
b) texture (standard deviation of gray-scale values)
c) perimeter
d) area
e) smoothness (local variation in radius lengths)
f) compactness (perimeter^2 / area - 1.0)
g) concavity (severity of concave portions of the contour)
h) concave points (number of concave portions of the contour)
i) symmetry
j) fractal dimension ("coastline approximation" - 1)
"""

wpbc_df = pd.read_csv("wpbc.data", header=None, names=column_headers)
wpbc_df.head(10)

0  1    2      3      4   ...      30      31       32   33  34
0  119513  N   31  18.02  27.60  ...  0.1170  0.2677  0.08113  5.0   5
1    8423  N   61  17.99  10.38  ...  0.2654  0.4601  0.11890  3.0   2
2  842517  N  116  21.37  17.44  ...  0.2032  0.4334  0.09067  2.5   0
3  843483  N  123  11.42  20.38  ...  0.2575  0.6638  0.17300  2.0   0
4  843584  R   27  20.29  14.34  ...  0.1625  0.2364  0.07678  3.5   0
5  843786  R   77  12.75  15.29  ...  0.1709  0.3485  0.11790  2.5   0
6  844359  N   60  18.98  19.61  ...  0.1922  0.2726  0.09581  1.5   ?
7  844582  R   77  13.71  20.83  ...  0.1556  0.3196  0.11510  4.0  10
8  844981  N  119  13.00  21.82  ...  0.2060  0.4378  0.10720  2.0   1
9  845010  N   76  12.46  24.04  ...  0.2210  0.4366  0.20750  6.0  20

[10 rows x 35 columns]

In [34]:

wpbc_df.tail(10)

0  1   2      3      4   ...       30      31       32   33  34
188   938413  N  15  12.53  30.98  ...  0.12590  0.3292  0.09406  3.0   7
189   939095  N   6  19.80  20.46  ...  0.18400  0.2382  0.06831  2.1   2
190   939426  N   8  19.96  27.41  ...  0.16380  0.2245  0.07842  4.0   0
191  9411286  N   3  19.22  27.18  ...  0.20990  0.3484  0.09774  0.8   1
192  9411300  N   3  14.72  25.26  ...  0.19740  0.3778  0.11320  1.7  21
193   942640  N  10  22.52  21.92  ...  0.09378  0.2061  0.05788  6.0   2
194   943471  N   8  15.44  31.18  ...  0.12860  0.2914  0.08024  1.5   0
195    94547  N  12  17.17  29.19  ...  0.09520  0.3362  0.06033  3.7   0
196   947204  R   3  21.42  22.84  ...  0.22150  0.2832  0.08981  3.0   ?
197   947489  N   6  16.70  28.13  ...  0.13170  0.3035  0.08036  3.5   0

[10 rows x 35 columns]

In [38]:
print(wpbc_df.describe())
print(wpbc_df.shape)
print(wpbc_df.info()) 

                 0           2           3   ...          31          32          33
count  1.980000e+02  198.000000  198.000000  ...  198.000000  198.000000  198.000000
mean   1.990469e+06   46.732323   17.412323  ...    0.323404    0.090828    2.847475
std    2.889025e+06   34.462870    3.161676  ...    0.075161    0.021172    1.937964
min    8.423000e+03    1.000000   10.950000  ...    0.156500    0.055040    0.400000
25%    8.557452e+05   14.000000   15.052500  ...    0.275950    0.076578    1.500000
50%    8.863390e+05   39.500000   17.290000  ...    0.310300    0.086890    2.500000
75%    9.279958e+05   72.750000   19.580000  ...    0.358800    0.101375    3.500000
max    9.411300e+06  125.000000   27.220000  ...    0.663800    0.207500   10.000000

[8 rows x 33 columns]
(198, 35)


## 3) Make Crosstabs of the Categorical Variables

Take whichever of the above datasets has more categorical variables and use crosstabs to tabulate the different instances of the categorical variables.


In [0]:
# Your Code Here

## 4) Explore the distributions of the variables of the dataset using:
- Histograms
- Scatterplots
- Density Plots

In [0]:
# Your Code Here

## 5) Create at least one visualization from a crosstab:

Remember that a crosstab is just a dataframe and can be manipulated in the same way by row index, column, index, or column/row/cell position.


In [0]:
# Your Code Here

## Stretch Goals 

The following additional study tasks are optional, they are intended to give you an opportunity to stretch yourself beyond the main requirements of the assignment. You can pick and choose from the below, you do not need to complete them in any particular order.

### - Practice Exploring other Datasets

### -  Try using the Seaborn plotting library's "Pairplot" functionality in order to explore all of the possible histograms and scatterplots of your dataset all at once:

[Seaborn Pairplot](https://seaborn.pydata.org/generated/seaborn.pairplot.html)

### - Turn some of the continuous variables into categorical variables by binning the values using:
- [pd.cut()](https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.cut.html)
- [pd.qcut()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html)
- <https://stackoverflow.com/questions/30211923/what-is-the-difference-between-pandas-qcut-and-pandas-cut>

And then use crosstabs to compare/visualize these binned variables against the other variables.


### - Other types and sources of data
Not all data comes in a nice single file - for example, image classification involves handling lots of image files. You still will probably want labels for them, so you may have tabular data in addition to the image blobs - and the images may be reduced in resolution and even fit in a regular csv as a bunch of numbers.

If you're interested in natural language processing and analyzing text, that is another example where, while it can be put in a csv, you may end up loading much larger raw data and generating features that can then be thought of in a more standard tabular fashion.

Overall you will in the course of learning data science deal with loading data in a variety of ways. Another common way to get data is from a database - most modern applications are backed by one or more databases, which you can query to get data to analyze. We'll cover this more in our data engineering unit.

How does data get in the database? Most applications generate logs - text files with lots and lots of records of each use of the application. Databases are often populated based on these files, but in some situations you may directly analyze log files. The usual way to do this is with command line (Unix) tools - command lines are intimidating, so don't expect to learn them all at once, but depending on your interests it can be useful to practice.

One last major source of data is APIs: https://github.com/toddmotto/public-apis

API stands for Application Programming Interface, and while originally meant e.g. the way an application interfaced with the GUI or other aspects of an operating system, now it largely refers to online services that let you query and retrieve data. You can essentially think of most of them as "somebody else's database" - you have (usually limited) access.

*Stretch goal* - research one of the above extended forms of data/data loading. See if you can get a basic example working in a notebook. Image, text, or (public) APIs are probably more tractable - databases are interesting, but there aren't many publicly accessible and they require a great deal of setup.